utils/benchmarking/eventing/timeseriesdb/timeseriesdb_test.go

```diff
@@ -0,0 +1,147 @@
+// MIT License
+//
+// Copyright (c) 2021 Mert Bora Alper and EASE lab
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+package main
+
+import (
+	"context"
+	"reflect"
+	"testing"
+	"time"
+
+	ctrdlog "github.com/containerd/containerd/log"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/ease-lab/vhive/utils/benchmarking/eventing/proto"
+)
+
+const SleepDuration = 200 * time.Millisecond
+
+func TestDefaultEndToEnd(t *testing.T) {
+	var server Server
+
+	// Parameters
+	// ================================================================
+	exDef := proto.ExperimentDefinition{CompletionEventDescriptors: []*proto.CompletionEventDescriptor{
+		{
+			AttrMatchers: map[string]string{
+				"type":   "greeting",
+				"source": "consumer",
+			},
+		},
+	}}
+	aInvokedOn := time.Now().UTC()
+	eventA1 := proto.Event{
+		VHiveMetadata: &proto.VHiveMetadata{Id: "A", InvokedOn: timestamppb.New(aInvokedOn)},
+		Attributes: map[string]string{
+			"id": "A",
+			"source": "producer",
+			"type": "greeting",
+		},
+		Data: []byte("some JSON data..."),
+	}
+	eventA2 := proto.Event{
+		VHiveMetadata: &proto.VHiveMetadata{Id: "A", InvokedOn: timestamppb.New(aInvokedOn)},
+		Attributes: map[string]string{
+			"id": "A",
+			"source": "consumer",
+			"type": "greeting",
+		},
+		Data: []byte("some JSON data..."),
```

Comment:
`eventA1` and `eventA2` differ at `.Attributes["source"]`. I cannot create a parent `eventA` to copy-and-modify as needed because because protobuf structs contain `sync.Mutex` which shall not be copied. I believe that this is a cleaner solution than its alternatives.